### PR TITLE
Test setup to verify ECMWF  produced forecast against our ETL output in database

### PR DESCRIPTION
**Pull request by mwalker-scottlogic**
### Description

This test is not fully automated yet - the process of running the main application and actually getting the ECMWF files is currently manual and requires another ticket to produce. However this can be run locally to check the divergence of cities for a specified forecast valid time from the ECMWF forecast. This is a method to indicate if we are calibrated with the current output. The tolerated divergence is set arbitrarily to 3% as present.

To run this test, you will need to download from the below links the list of CNN countries, and the desired forecast, and save the files within `air-quality-backend`. You then specify their path within `air-quality-backend/system_tests/forecast_calibration_tests.py`.

You then need to specify a matching base date for the forecast and a desired valid time to check against.

-  List of stations (https://myftp.ecmwf.int/view/public/cams/products/cams_global_forecast/surface_concentrations/CAMS_locations_V1.csv) • 
- Latest air quality forecasts for the list of the CNN cities in CSV format (https://myftp.ecmwf.int/files/public/cams/products/cams_global_forecast/surface_concentrations/)

`air-quality-backend/system_tests/cities_data.py`

- created two datasets
- all cities & a subset that have longitudes closest to -180, 0 and 180

`air-quality-backend/system_tests/forecast_calibration_tests.py`

- main test file
- tests are parameterised to run against desired dataset
- tests assert if database data for )3, NO2, PM10 and PM2_5 are within the desired tolerance of the ECMWF forecast

`air-quality-backend/system_tests/utils/helper_methods.py`

- amended functionality of the `get_database_data()` function and added new helper methods for the tests

Verification

- Output has been checked manually, an example below shows the test identifying that the only city above 2% divergence at a given base time and time slot for Ozone was Bogota

image

image

### CODE REVIEW

File: air-quality-backend/system_tests/utils/helper_methods.py
Comment (amehta-scottlogic) [inline air-quality-backend/system_tests/utils/helper_methods.py:1]: The code in here is really clean and understandable - looks good.

I would rename this file to be 'forecast_utilities' or 'cams_utilities' to be more specific to what it includes. I imagine then you can have a separate file for other aspects such as in_situ, api etc

File: air-quality-backend/system_tests/cities_data.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+all_cities = [
```
[the remark this answers is not in the document]
Comment (mwalker-scottlogic) [reply air-quality-backend/system_tests/cities_data.py:1]: Done as suggested

File: air-quality-backend/system_tests/forecast_calibration_tests.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        filter(
+            lambda x: x["forecast_base_time"] == test_forecast_base_time
+            and x["name"] == test_city
+            and x["forecast_valid_time"] == test_forecast_valid_time,
+            database_all_data,
+        )
+    )
+
+    database_no2_value = database_record_for_city_and_valid_time[0]["no2_value"]
+    ecmwf_forecast_no2_value = ecmwf_record_for_city_and_valid_time[0]["NO2"]
```
[the remark this answers is not in the document]
Comment (mwalker-scottlogic) [reply air-quality-backend/system_tests/forecast_calibration_tests.py:94]: Hi, I've made some changes based on the above, but if this doesn't match your intentions please do let me know

**Comment by mwalker-scottlogic** [comment]
Added changes, retested at a 1% threshold and compared to doing the check manually and the results matched

**Review by amehta-scottlogic — APPROVED** [review]

**Review by amehta-scottlogic — APPROVED** [review]

**Pull request merged**